tests/linearizability/watch.go

```diff
@@ -85,61 +85,94 @@ func validateWatchResponses(t *testing.T, responses [][]watchResponse, expectPro
+	}
+}
 
+func validateEventsWithRevisionLowerThanProgressNotifyWereDelivered(t *testing.T, responses []watchResponse) {
+	var lastProgressNotifyRevision int64 = 0
+	for _, resp := range responses {
 		for _, event := range resp.Events {
-			rk := revisionKey{key: string(event.Kv.Key), revision: event.Kv.ModRevision}
-			if _, found := uniqueOperations[rk]; found {
-				t.Errorf("BROKE: Within the:wq same revision key can only be modified once. Suspecting duplicate watch event. key: %q, revision: %d", rk.key, rk.revision)
-			}
-			uniqueOperations[rk] = struct{}{}
 			if event.Kv.ModRevision <= lastProgressNotifyRevision {
-				t.Errorf("BROKE: etcd will not send progress notification to a watcher until it has synced all events. So a watcher will never receive an event with a lower `ModRevision` than the last progressNotification's revision, eventRevision: %d, progressNotifyRevision: %d", event.Kv.ModRevision, lastProgressNotifyRevision)
+				t.Errorf("BROKE: Events with revision lower than progress notify were delivered, eventRevision: %d, progressNotifyRevision: %d", event.Kv.ModRevision, lastProgressNotifyRevision)
 			}
+		}
+		if resp.IsProgressNotify() {
+			lastProgressNotifyRevision = resp.Header.Revision
+		}
+	}
+}
+
+func validateEventRevisionIsContinuousAndNonDecreasing(t *testing.T, responses []watchResponse) {
+	var lastEventModRevision int64 = 1
+	for _, resp := range responses {
+		for _, event := range resp.Events {
 			if event.Kv.ModRevision != lastEventModRevision && event.Kv.ModRevision != lastEventModRevision+1 {
-				t.Errorf("Event mod revision should stay the same or increment by one, last: %d, current: %d", lastEventModRevision, event.Kv.ModRevision)
+				t.Errorf("Event revision should stay the same or increment by one, last: %d, current: %d", lastEventModRevision, event.Kv.ModRevision)
 			}
 			lastEventModRevision = event.Kv.ModRevision
 		}
+	}
+}
 
-		if resp.Header.Revision < lastEventModRevision {
-			t.Errorf("Event revision should never exceed the server's revision, lastEventRevision: %d, resp.Header.Revision: %d",
-				lastEventModRevision, resp.Header.Revision)
+func validateWatchEventsAreUnique(t *testing.T, responses []watchResponse) {
+	type revisionKey struct {
+		revision int64
+		key      string
+	}
+	uniqueOperations := map[revisionKey]struct{}{}
+	for _, resp := range responses {
+		for _, event := range resp.Events {
+			rk := revisionKey{key: string(event.Kv.Key), revision: event.Kv.ModRevision}
+			if _, found := uniqueOperations[rk]; found {
+				t.Errorf("BROKE: Watch events are unique, key: %q, revision: %d", rk.key, rk.revision)
+			}
+			uniqueOperations[rk] = struct{}{}
 		}
+	}
+}
 
-		if resp.IsProgressNotify() {
-			lastProgressNotifyRevision = resp.Header.Revision
+func validateHeaderRevisionGreaterOrEqualToEventRevision(t *testing.T, responses []watchResponse) {
+	for _, resp := range responses {
+		for _, event := range resp.Events {
+			if resp.Header.Revision < event.Kv.ModRevision {
```

Comment:
We have already verified that the event revision is strictly incremental in `validateEventRevisionIsContinuousAndNonDecreasing`, please just compare headerRevision with the last event revision if present.

Reply:
Want to keep the validators independent.

Reply:
OK, this isn't a big deal. Feel free to keep it as it's.